### PR TITLE
cof editor (resolves #82)

### LINKED ISSUE #82: Add editor features to COF viewer
Depends on:
* https://github.com/OpenDiablo2/OpenDiablo2/issues/909
* #71

### PR DESCRIPTION
**Pull request by gucio321**
Hi there,
here are editor features for cof viewer.

solves #82

### CODE REVIEW

File: hswidget/cofeditor.go [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+// COFEditor contains data necessary do edit cof file
+type COFEditor struct {
+	newCofLayer       *d2cof.CofLayer
+	cof               *d2cof.COF
+	id                string
+	upArrowTexture    *giu.Texture
+	downArrowTexture  *giu.Texture
+	leftArrowTexture  *giu.Texture
+	rightArrowTexture *giu.Texture
+}
```
Comment (gucio321) [inline hswidget/cofeditor.go]: what do you think, maybe I should rename COFViewerWidget into COFEditorWIdget and place this code in COFEditorState?

File: hscommon/hsenum/cof_editor_state.go [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+// COFEditorState represents cof editor's states
+type COFEditorState int
+
+// cof editor's states
+const (
+	COFEditorStateViewer COFEditorState = iota
+	COFEditorStateAddLayer
+	COFEditorStateAddDirection
+	COFEditorStateConfirm
+)
```
Comment (gravestench) [inline hscommon/hsenum/cof_editor_state.go]: I dont think we need to put _every_ enumeration into `hsenum`, only the enums we want to re-use. Only the COF editor/viewer uses these states, I would imagine that these would be defined in ~~`hswindow/hseditor/hscofeditor/`~~ `hswidget/` and not be exported.

File: hscommon/hsenum/draw_effect_names.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+	case d2enum.DrawEffectMod2XTrans:
+		effect = "Mod2XTrans"
+	case d2enum.DrawEffectMod2X:
+		effect = "Mod2X"
+	case d2enum.DrawEffectNone:
+		effect = stringNone
+	}
+
+	return effect
+}
```
Comment (gravestench) [inline hscommon/hsenum/draw_effect_names.go:37]: I think we should try to avoid large switch statements like these when possible.
```golang
func GetDrawEffectName(effect d2enum.DrawEffect) string {
	const (
		strNone    = "None"
		strUnknown = "Unknown"
	)

	effectNames := map[d2enum.DrawEffect]string{
		d2enum.DrawEffectPctTransparency25: "25% alpha",
		d2enum.DrawEffectPctTransparency50: "50% alpha",
		d2enum.DrawEffectPctTransparency75: "75% alpha",
		d2enum.DrawEffectModulate:          "Modulate",
		d2enum.DrawEffectBurn:              "Burn",
		d2enum.DrawEffectNormal:            "Normal",
		d2enum.DrawEffectMod2XTrans:        "Mod2XTrans",
		d2enum.DrawEffectMod2X:             "Mod2X",
		d2enum.DrawEffectNone:              strNone,
	}

	effectName, found := effectNames[effect]
	if !found {
		return strUnknown
	}

	return effectName
}
```

Comment (gucio321) [reply hscommon/hsenum/draw_effect_names.go:37]: ok, it was 1:1 copy of existing in hswidget/cofviewer.go.
I'll rewrite

File: hscommon/hsenum/weapon_class_string.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+		weapon = "Left Swing Right Swing"
+	case d2enum.WeaponClassLeftSwingRightThrust:
+		weapon = "Left Swing Right Thrust"
+	case d2enum.WeaponClassOneHandToHand:
+		weapon = "One Hand To Hand"
+	case d2enum.WeaponClassTwoHandToHand:
+		weapon = "Two Hand To Hand"
+	}
+
+	return weapon
```
Comment (gravestench) [inline hscommon/hsenum/weapon_class_string.go:45]: Same here; we can avoid the need for a nolint directive if we use a map:
```golang
func GetWeaponClassString(weaponClass d2enum.WeaponClass) string {
	const (
		strNone    = "None"
		strUnknown = "Unknown"
	)

	weaponClassNames := map[d2enum.WeaponClass]string{
		d2enum.WeaponClassNone:                 strNone,
		d2enum.WeaponClassHandToHand:           "Hand To Hand",
		d2enum.WeaponClassBow:                  "Bow",
		d2enum.WeaponClassOneHandSwing:         "One Hand Swing",
		d2enum.WeaponClassOneHandThrust:        "One Hand Thrust",
		d2enum.WeaponClassStaff:                "Staff",
		d2enum.WeaponClassTwoHandSwing:         "Two Hand Swing",
		d2enum.WeaponClassTwoHandThrust:        "Two Hand Thrust",
		d2enum.WeaponClassCrossbow:             "Crossbow",
		d2enum.WeaponClassLeftJabRightSwing:    "Left Jab Right Swing",
		d2enum.WeaponClassLeftJabRightThrust:   "Left Jab Right Thrust",
		d2enum.WeaponClassLeftSwingRightSwing:  "Left Swing Right Swing",
		d2enum.WeaponClassLeftSwingRightThrust: "Left Swing Right Thrust",
		d2enum.WeaponClassOneHandToHand:        "One Hand To Hand",
		d2enum.WeaponClassTwoHandToHand:        "Two Hand To Hand",
	}

	weaponClassName, found := weaponClassNames[weaponClass]
	if !found {
		return strUnknown
	}

	return weaponClassName
}
```

Comment (gravestench) [reply hscommon/hsenum/weapon_class_string.go:45]: honestly, I think it might be worthwhile to add methods to our enum types to avoid having to do this:
```golang
func (wc WeaponClass) DescriptiveName() string {
	const (
		strNone    = "None"
		strUnknown = "Unknown"
	)
	
	descriptions := map[WeaponClass]string{
		WeaponClassNone:                 strNone,
		WeaponClassHandToHand:           "Hand To Hand",
		WeaponClassBow:                  "Bow",
		WeaponClassOneHandSwing:         "One Hand Swing",
		WeaponClassOneHandThrust:        "One Hand Thrust",
		WeaponClassStaff:                "Staff",
		WeaponClassTwoHandSwing:         "Two Hand Swing",
		WeaponClassTwoHandThrust:        "Two Hand Thrust",
		WeaponClassCrossbow:             "Crossbow",
		WeaponClassLeftJabRightSwing:    "Left Jab Right Swing",
		WeaponClassLeftJabRightThrust:   "Left Jab Right Thrust",
		WeaponClassLeftSwingRightSwing:  "Left Swing Right Swing",
		WeaponClassLeftSwingRightThrust: "Left Swing Right Thrust",
		WeaponClassOneHandToHand:        "One Hand To Hand",
		WeaponClassTwoHandToHand:        "Two Hand To Hand",
	}

	description, found := descriptions[wc]
	if !found {
		return strUnknown
	}
	
	return description
}
```
This way, we can just use `wc.DescriptiveName()` wherever we need, without having to make new maps or remember which phrozen keep forum post to look at.

However, this edit would need to be done in the `OpenDiablo2` repo

File: hscommon/hsenum/draw_effect_names.go (hunk 3 above)
Comment (gravestench) [reply hscommon/hsenum/draw_effect_names.go:37]: no worries, i'm likely to blame for that haha

File: hswindow/hseditor/hscofeditor/cof_editor.go [hunk 5]
```diff
@@ -21,11 +21,12 @@ var _ hscommon.EditorWindow = &COFEditor{}
 // COFEditor represents a cof editor
 type COFEditor struct {
 	*hseditor.Editor
-	cof *d2cof.COF
+	cof       *d2cof.COF
+	cofEditor *hswidget.COFEditor
```
Comment (gucio321) [inline hswindow/hseditor/hscofeditor/cof_editor.go]: maybe it should be in `hswidget.COFViewerState`?

**Comment by gravestench** [comment]
disregard my previous comments, i'll take care of the long switch statements in another PR

File: hswidget/cof_editor_widget.go [hunk 6]
```diff
@@ -0,0 +1,529 @@
...
+		if int32(n) != index {
+			newPriority = append(newPriority, i)
+		}
+	}
+
+	p.cof.Priority = newPriority
+}
+
+// CreateNewLayer starts add-cof-layer dialog
+func (p *COFWidget) CreateNewLayer() {
```
Comment (gravestench) [inline hswidget/cof_editor_widget.go:522]: we can probably break up this file further in another PR, but this looks good for now

**Review by gravestench — APPROVED** [review]

**Comment by gravestench** [comment]
Resolves #82
